### DISCUSSION TITLE
Is there a correct way to calculate a popup menu height *before* displaying it?

### DISCUSSION THREAD

**Discussion by chezbel** [fltk/fltk#1011] [General]
Is there an API function that calculates the height of a menu before 
it is displayed? I want to prevent a popup menu from extending below 
the bottom of a window, which is easy if the height of a menu is known.

I'd even be happy if the API took a "long" time such as doing all the 
normal work needed to show a menu, but threw it away at the last second
and returned the height and width. 

On the other hand if there is a correct formula for calculating the height
that would be enough.

Anyway, I don't *need* this as I think I finally got something to work,
but part of the calculation is based on trial and error so I'm not sure
if it's really correct. I'm currently doing the following to calculate
menuY, which is the Y position to use for Fl_Menu_Button that causes
the menu to be at the very bottom of the window when popup() is called.
```
1) int menuHeight = menuItems * fontHeight;
2) menuHeight += (menuItems * 4);
3) menuY = window->y() + windown->h() - 1 - menuHeight - 2;
```
#2 I guess this is for spacing between menu rows? I'm not sure...
#3 I'm not sure why I needed the "-2", maybe related to borders?

Again, this isn't super important as I seem to have something working
pretty well.

**Comment by Albrecht-S** [comment]
@chezbel This is an interesting question, and I believe the general answer must be _"no, unfortunately not"_.

There are some issues that may interfere with your calculation and - even more important - with the positioning of the menu.

- Item `#2` above: yes, this should correspond to [Fl::menu_linespacing()](https://www.fltk.org/doc-1.4/classFl.html#ab2d29df12fea1b116d70e237eb471798) whose default value is 4. I suggest to use the global setting instead of the literal value `4`. You can change this value with `Fl::menu_linespacing(int value)` to test if it works for you.
- Item `#3` above: you are likely right, but the border width and height is variable per boxtype, and you can (IIRC) set the boxtype of a menu (button) or something like that, I'm not sure about the details. The corresponding method to find the right value would probably be something like `Fl::box_dy(boxtype)`.
- Font sizes may vary across platforms, and thus the `fontHeight` (whatever you are using for this value) may include a pixel or more (whitespace, not real printing area) on different platforms.
- Last but not least, using `window->y()` to position another window relative to `this` window doesn't work on the `Wayland` platform because Wayland hides the real window position of top windows on the desktop (screen). FLTK menus can be positioned relatively to their parent window, but AFAICT there is no method available to the user program to do something similar. Our Wayland expert @ManoloFLTK may know more about the details.

You wrote:
> I want to prevent a popup menu from extending below the bottom of a window, which is easy if the height of a menu is known.

Do you really mean the bottom of the window, or would it be sufficient to keep it inside the screen? The latter is guaranteed by FLTK anyway.

**Comment by chezbel** [comment]
Thanks for the detailed reply, it included some generally useful info that I didn't know before
- For item 2 above, using Fl::menu_linespacing() instead of "4" works, even when Fl::menu_linspacing() was set to a few different values, e.g. 6 and 10.
- For item 3 above, using Fl::box_dy(boxtype) instead of "2" works. I tested with FL_FLAT_BOX (dy is 0), FL_UP_BOX (dy is 2) and FL_THIN_UP_BOX (dy is 1).

That's great because my calculation no longer includes mysterious hard coded values and it is more correct.

In my previous example code fontHeight comes directly from fl_height(). I think it doesn't matter if that includes extra pixels for whitespace, unless the menu drawing code is using a row height different from fl_height().

So I think the only big issue is Wayland and currently I can ignore that as I'm only using Ubuntu with X11. In the future who knows...

To answer your question, yes I really did mean the bottom of the window. My app is a text editor and I wanted to constrain the menu to being inside the text editing area without extending downward and covering part of the status line and/or file tabs. Of course it isn't *necessary* to do this, but it's a design I like and wanted to implement and it seems to be working ok now on my platform.

Off topic: After posting both here and in fltk.general, both work and are reasonable. However, I like the layout of GitHub better and the message Preview is really nice. Also, GitHub doesn't require logging in to Google, which I avoid doing on my PC unless absolutely necessary.

**Reply by ManoloFLTK** [reply]
> So I think the only big issue is Wayland ...

Menu windows under Wayland automatically stay inside the screen, unless they are higher than the screen itself, which is probably not your situation.
If user code makes sure under X11 a menu window is inside the bounds of an Fl_Window and if that Fl_Window itself is entirely inside the screen, this code will produce exactly the same result if run under Wayland.

**Comment by MatthiasWM** [comment]
Pulldown and popup menus height is relatively easy to calculate. All menu items are always the same height, even if the font sizes differ or there is a divider line somewhere. The item height is the maximum height of all visible menu items plus `Fl::menu_linespacing()`. If the menu item has no image, it's simply `fl_height()`. If it does, it's the maximum of the image height and `fl_height()` of that menu item font. There are a few other styles that can usually be ignored, and I am assuming you use no newline characters in the label.

Use `Fl_Menu_Item::next(int n)` to iterate through _visible_ items.

Multiply the itemheight with the number of visible items, subtract `Fl::menu_linespacing()` once for the lowest item, and add the box height `Fl::box_dh(b)`. The box type is either set in the corresponding menu button or `FL_UP_BOX`, but usually less than 8 pixels, so no huge deal.
